presto-main/src/main/java/com/facebook/presto/operator/scalar/BitwiseFunctions.java

```diff
@@ -75,4 +83,234 @@ public static long bitwiseXor(@SqlType(StandardTypes.BIGINT) long left, @SqlType
+    }
+
+    @Description("arithmetic shift right operation")
+    @ScalarFunction
+    @SqlType(StandardTypes.BIGINT)
+    public static long bitwiseArithmeticShiftRight(@SqlType(StandardTypes.BIGINT) long number, @SqlType(StandardTypes.BIGINT) long shift)
+    {
+        if (shift < 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Specified shift must be positive");
+        }
+
+        return number >> shift;
+    }
+
+    @Description("bitwise left shift")
+    @ScalarFunction("bitwise_left_shift")
+    @SqlType(StandardTypes.TINYINT)
+    public static long bitwiseLeftShiftTinyint(@SqlType(StandardTypes.TINYINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= MAX_BITS) {
+            return 0L;
+        }
+        long shifted = (value << shift);
+        return preserveSign(shifted, TINYINT_MASK, TINYINT_SIGNED_BIT);
+    }
+
+    @Description("bitwise left shift")
+    @ScalarFunction("bitwise_left_shift")
+    @SqlType(StandardTypes.SMALLINT)
+    public static long bitwiseLeftShiftSmallint(@SqlType(StandardTypes.SMALLINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= MAX_BITS) {
+            return 0L;
+        }
+        long shifted = (value << shift);
+        return preserveSign(shifted, SMALLINT_MASK, SMALLINT_SIGNED_BIT);
+    }
+
+    @Description("bitwise left shift")
+    @ScalarFunction("bitwise_left_shift")
+    @SqlType(StandardTypes.INTEGER)
+    public static long bitwiseLeftShiftInteger(@SqlType(StandardTypes.INTEGER) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= MAX_BITS) {
+            return 0L;
+        }
+        long shifted = (value << shift);
+        return preserveSign(shifted, INTEGER_MASK, INTEGER_SIGNED_BIT);
+    }
+
+    @Description("bitwise left shift")
+    @ScalarFunction("bitwise_left_shift")
+    @SqlType(StandardTypes.BIGINT)
+    public static long bitwiseLeftShiftBigint(@SqlType(StandardTypes.BIGINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= MAX_BITS) {
+            return 0L;
+        }
+        return value << shift;
+    }
+
+    private static long preserveSign(long shiftedValue, long mask, long signedBit)
+    {
+        if ((shiftedValue & signedBit) != 0) {
+            // Preserve the sign in 2's complement format
+            return shiftedValue | ~mask;
+        }
+
+        return shiftedValue & mask;
+    }
+
+    @Description("bitwise logical right shift")
+    @ScalarFunction("bitwise_right_shift")
+    @SqlType(StandardTypes.TINYINT)
+    public static long bitwiseRightShiftTinyint(@SqlType(StandardTypes.TINYINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= MAX_BITS) {
+            return 0L;
+        }
+        if (shift == 0) {
+            return value;
+        }
+        return (value & TINYINT_MASK) >>> shift;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 205:211, 219:225, 233:239.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.